quine-core/src/main/scala/com/thatdot/quine/graph/NodeActor.scala

```diff
@@ -147,38 +209,12 @@ private[graph] class NodeActor(
     events: Seq[NodeChangeEvent],
     atTimeOverride: Option[EventTime] = None
   ): Future[Done.type] =
-    if (atTime.isDefined) Future.failed(IllegalHistoricalUpdate(events, qid, atTime.get))
+    if (atTime.isDefined)
+      Future.failed(IllegalHistoricalUpdate(events, qid, atTime.get))
     else if (atTimeOverride.isDefined && events.size > 1)
       Future.failed(IllegalTimeOverride(events, qid, atTimeOverride.get))
     else {
-      val dedupedEffectingEvents: Seq[NodeChangeEvent.WithTime] = {
-        if (events.isEmpty) Seq.empty
-        else if (events.size == 1 && hasEffect(events.head))
-          Seq(NodeChangeEvent.WithTime(events.head, atTimeOverride.getOrElse(nextEventTime())))
-        else {
-          /* This process reverses the events, considering only the last event per property/edge/etc. and keeps the
-           * event if it has an effect. If multiple events would affect the same value (e.g. have the same property key),
-           * but would result in no change when applied in order, then no change at all will be applied. e.g. if a
-           * property exists, and these events would remove it and set it back to its same value, then no change to the
-           * property will be recorded at all. Original event order is maintained. */
-          var es: Set[HalfEdge] = Set.empty
-          var ps: Set[Symbol] = Set.empty
-          var ft: Option[QuineId] = None
-          var mih: Set[QuineId] = Set.empty
-          events.reverse
-            .filter {
-              case e @ EdgeAdded(ha) => if (es.contains(ha)) false else { es += ha; hasEffect(e) }
-              case e @ EdgeRemoved(ha) => if (es.contains(ha)) false else { es += ha; hasEffect(e) }
-              case e @ PropertySet(k, _) => if (ps.contains(k)) false else { ps += k; hasEffect(e) }
-              case e @ PropertyRemoved(k, _) => if (ps.contains(k)) false else { ps += k; hasEffect(e) }
-              case e @ MergedIntoOther(id) => if (ft.isDefined) false else { ft = Some(id); hasEffect(e) }
-              case e @ MergedHere(o) => if (mih.contains(o)) false else { mih += o; hasEffect(e) }
-            }
-            .reverse
-            .map(e => WithTime(e, atTimeOverride.getOrElse(nextEventTime())))
-          // TODO: It should be possible to do all this in only two passes over the collection with no reverses.
```

Comment:
I think the reasoning about how this change would affect the performance needs to distinguish between computational complexity and heap complexity.

If we look at the original (and assumpting that events: Seq[NodeChangedEvent] is implemented as a List):
- The each of the reverse operations do an allocation for each cell in the list.
- Building the Sets will do an allocation for each element.
- The filter is going to a heap allocation for each element of the list.

Overall, that is about 5 allocations for each element events. Even if it is short-lived heap froth, it still seems like a lot of allocations.

The proposed essentially removes the allocations, except for the ArrayBuffer (two heap allocations). The cost is the search through the later events, which is quadratic. If events is small, this is not going to be significant. The question is at what point would the length of events be signficant?

I think that determining whether this change would result in better performance would require:

1.  Finding the distribution of lengths of events we expect to see, and perhaps to find a maximum expected upper bound for the length.
2.  Do some performance tests (e.g., with generated data) to see how each of the cases performs.

I don't really know #1 at this point, so I'd say that this change is not justified. It might be useful, but I wouldn't do it without actually doing the work to justify it.